backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/FunctionReferenceLowering.kt

```diff
@@ -292,19 +293,20 @@ internal class FunctionReferenceLowering(val context: Context): FileLoweringPass
                     suspendFunctionClass = symbols.suspendFunctionN(numberOfParameters - 1).owner
                     val suspendFunctionClassTypeParameters = functionParameterTypes.dropLast(1) +
                             (lastParameterType.arguments.single().typeOrNull ?: irBuiltIns.anyNType)
-                    superTypes += suspendFunctionClass.symbol.typeWith(suspendFunctionClassTypeParameters)
+                    if (samSuperType == null)
+                        superTypes += suspendFunctionClass.symbol.typeWith(suspendFunctionClassTypeParameters)
                 }
             }
 
             val constructor = buildConstructor()
             val functionInvoke = if (isKSuspendFunction)
                 null
             else
-                buildInvokeMethod(functionClass.getInvokeFunction())
+                buildInvokeMethod(functionClass.getInvokeFunction(), putToOverriddenSymbols = samSuperType == null)
```

Comment:
And now this probably just builds redundant method (e.g. if SAM has extension receiver).

I guess this code is quite complicated, so optimizing it would make it even worse.
On the other hand, building a specialized implementation for SAM might help to make it more precise and simplify the implementation of both (SAM and regular function references/lambdas). WDYT?

Reply:
>building a specialized implementation for SAM might help to make it more precise and simplify the implementation of both
Makes sense, I don't remember why I hadn't implemented it that way in the first place.